File: golang/231_is_power_of_two/is_power_of_two.go
/*
 * @Date: 2021-05-30 09:22:59
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-30 09:30:05
 */

package main

func isPowerOfTwo_1(n int) bool {
	return n > 0 && n&(n-1) == 0
}

func isPowerOfTwo_2(n int) bool {
	return n > 0 && n&-n == n
}
func isPowerOfTwo_3(n int) bool {
	const big = 1 << 30
	return n > 0 && big%n == 0
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed!")
		}
	}
	assert(isPowerOfTwo_1(1))
	assert(isPowerOfTwo_2(1))
	assert(isPowerOfTwo_3(1))
	assert(isPowerOfTwo_1(16))
	assert(isPowerOfTwo_2(16))
	assert(isPowerOfTwo_3(16))
	assert(!isPowerOfTwo_1(3))
	assert(!isPowerOfTwo_2(3))
	assert(!isPowerOfTwo_3(3))
	assert(isPowerOfTwo_1(4))
	assert(isPowerOfTwo_2(4))
	assert(isPowerOfTwo_3(4))
	assert(!isPowerOfTwo_1(5))
	assert(!isPowerOfTwo_2(5))
	assert(!isPowerOfTwo_3(5))
}

File: rust/231_is_power_of_two/is_power_of_two.rs
/*
 * @Date: 2021-05-30 09:22:54
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-30 09:36:44
 */

fn is_power_of_two_1(n: i32) -> bool {
    n > 0 && n & (n - 1) == 0
}
fn is_power_of_two_2(n: i32) -> bool {
    n > 0 && (n & -n) == n
}
fn is_power_of_two_3(n: i32) -> bool {
    const BIG: i32 = 1 << 30;
    n > 0 && BIG % n == 0
}

fn main() {
    assert!(is_power_of_two_1(1));
    assert!(is_power_of_two_2(1));
    assert!(is_power_of_two_3(1));
    assert!(is_power_of_two_1(16));
    assert!(is_power_of_two_2(16));
    assert!(is_power_of_two_3(16));
    assert!(!is_power_of_two_1(3));
    assert!(!is_power_of_two_2(3));
    assert!(!is_power_of_two_3(3));
    assert!(is_power_of_two_1(4));
    assert!(is_power_of_two_2(4));
    assert!(is_power_of_two_3(4));
    assert!(!is_power_of_two_1(5));
    assert!(!is_power_of_two_2(5));
    assert!(!is_power_of_two_3(5));
}
